### PR TITLE
add after_train_iter

### PR DESCRIPTION
**Pull request by hnyu**
Added a hook `after_train_iter` to be called after each training iteration. The next PR would be to make non-RL algorithms also make use of most of `train_from_unroll` and `train_from_replay_buffer` (put common code into helper functions). 

A major change is that currently we unroll for off-policy algorithms with `torch.no_grad()`. Now that part also has grads for optionally training on-policy algorithms in `after_train_iter`. 

Example usage (for now assume that `self._predictor` derives from `OnPolicyAlgorithm` or `OffPolicyAlgorithm` in order to use the two functions):

```python

# in an off-policy agent.py:
def after_train_iter(self, training_info):
     self._predictor.train_from_unroll(getattr(training_info, "predictor"))
```
or 
```python
# in an off-policy agent.py:
def after_train_iter(self, training_info):
     self._predictor.train_from_replay_buffer() # has its own replay buffer
```

### CODE REVIEW

**Comment by hnyu** [comment]
@le-horizon  My last commit adds a comment about removing the tmp solution of detaching in RLAlgorithm.observe() after you fix the detach issue

**Comment by le-horizon** [comment]
> @le-horizon My last commit adds a comment about removing the tmp solution of detaching in RLAlgorithm.observe() after you fix the detach issue

Roger!

File: alf/algorithms/algorithm.py [hunk 1]
```diff
@@ -737,17 +737,41 @@ def update_with_gradient(self, loss_info, valid_masks=None, weight=1.0):
...
 
         Args:
-            training_info (nested Tensor): information collected for training.
+            training_info (TrainingInfo): information collected for training.
                 It is batched from each ``info`` returned by ``rollout_step()`` or
                 ``train_step()``.
         """
         pass
 
+    def after_train_iter(self, training_info=None):
```
Comment (emailweixu) [inline alf/algorithms/algorithm.py:751]: Can you also update the comments for RLAlgorithm about all the major hooks?

File: alf/algorithms/off_policy_algorithm.py [hunk 2]
```diff
@@ -105,18 +101,41 @@ def _train_iter_off_policy(self):
...
+                training_info = self.unroll(config.unroll_length)
+                self.summarize_rollout(training_info)
+                self.summarize_metrics()
+                if self._exp_replayer.total_size >= config.initial_collect_steps:
+                    break
+
+        steps = self.train_from_replay_buffer(True)
+
+        with record_time("time/after_train_iter"):
+            # skip the first iteration as it can unroll quite long
```
Comment (emailweixu) [inline alf/algorithms/off_policy_algorithm.py]: I don't understand the reason to skip.

File: alf/algorithms/off_policy_algorithm.py [hunk 3]
```diff
@@ -105,18 +101,41 @@ def _train_iter_off_policy(self):
         """User may override this for their own training procedure."""
         config: TrainerConfig = self._config
 
+        first_iter = (alf.summary.get_global_counter() == 0)
+
         if not config.update_counter_every_mini_batch:
             alf.summary.get_global_counter().add_(1)
 
-        with torch.no_grad():
```
Comment (emailweixu) [inline alf/algorithms/off_policy_algorithm.py:111]: As a note, it will cost more memory usage without "with no_grad()".

File: alf/algorithms/algorithm.py [hunk 4]
```diff
@@ -737,17 +737,41 @@ def update_with_gradient(self, loss_info, valid_masks=None, weight=1.0):
...
+        is always only one gradient update in each training iteration. Otherwise
+        it's less frequently called than ``after_update``.
+
+        Args:
+            training_info (TrainingInfo): information collected during ``unroll()``.
+                Note that it won't contain the field ``rollout_info`` because this
+                is the info collected just from the unroll but not from a replay
+                buffer. So if ``training_info`` is used, make sure you are doing
+                on-policy training in this function; if it's None, then only
+                off-policy training is allowed.
```
Comment (emailweixu) [inline alf/algorithms/algorithm.py]: comment when training_info is None.

File: alf/algorithms/off_policy_algorithm.py (hunk 2 above)
Comment (hnyu) [reply alf/algorithms/off_policy_algorithm.py]: Because we have init_collect steps. It might be thousands of steps. I've chopped init_collect_steps into multiples of unroll_length.

File: alf/algorithms/algorithm.py (hunk 1 above)
Comment (hnyu) [reply alf/algorithms/algorithm.py:751]: updated

File: alf/algorithms/off_policy_algorithm.py (hunk 3 above)
Comment (hnyu) [reply alf/algorithms/off_policy_algorithm.py:111]: Yeah I know that. Potentially we could optimize this part by setting a flag indicating whether we should unroll with grads or not depending on the user's need for training sub-algorithms. But if `unroll_length` is small then it's probably not a big issue.

File: alf/algorithms/sac_algorithm_test.py [hunk 5]
```diff
@@ -87,7 +87,7 @@ def test_sac_algorithm(self):
             name="MySAC")
 
         eval_env.reset()
-        for i in range(200):
+        for i in range(700):
```
Comment (Haichao-Zhang) [inline alf/algorithms/sac_algorithm_test.py:90]: what's the reason for increasing this number?

File: alf/algorithms/agent_helpers.py [hunk 6]
```diff
@@ -23,126 +23,164 @@
...
+
+    def state_specs(self):
+        """Return the state specs collected from child algorithms."""
+        return dict(
+            train_state_spec=self._train_state_spec,
+            rollout_state_spec=self._rollout_state_spec,
+            predict_state_spec=self._predict_state_spec)
+
+    @staticmethod
+    def accumulate_algortihm_rewards(rewards, weights, names, summary_prefix,
```
Comment (Haichao-Zhang) [inline alf/algorithms/agent_helpers.py]: ```accumulate_algortihm_rewards``` -> ```accumulate_algorithm_rewards```; and all the places calling this method

**Review by Haichao-Zhang — CHANGES_REQUESTED** [review]

File: alf/algorithms/sac_algorithm_test.py (hunk 5 above)
Comment (hnyu) [reply alf/algorithms/sac_algorithm_test.py:90]: Now init_collect_steps is divided into multiple iterations

Comment (Haichao-Zhang) [reply alf/algorithms/sac_algorithm_test.py:90]: I see. Do you think counting the ```initial_collection_steps``` into ```num_of_iters``` is a bit confusing?
For example, in this case, the first 500 training iterations are essentially for data collection and not doing any training, which might not be very clear when configuring gins.

Comment (hnyu) [reply alf/algorithms/sac_algorithm_test.py:90]: I don't see another better way. I think we always choose a large enough num_iterations in a gin file. The point is, initial_collection_steps has to be divided if grads are recorded for a flexible training of sub-algorithms.  Otherwise the graph will be really huge.

Comment (emailweixu) [reply alf/algorithms/sac_algorithm_test.py:90]: Perhaps update the comment about `num_iterations` at TrainerConfig?

Comment (hnyu) [reply alf/algorithms/sac_algorithm_test.py:90]: updated

**Review by Haichao-Zhang — APPROVED** [review]

**Review by emailweixu — APPROVED** [review]
